Update wasmtime dependency from 18 to 21

FrankReh opened the pull request:
Update the wasmtime dependency version from 18 to 21.

Also add type checks for the wasmtime/component.

Removed the `ty` method from the `WasmValue` trait.

Also add type checks for the value::{Value, Type} as the `ty` method is no longer available for building up a tree of types from a tree of values.

FrankReh committed:
bump wasmtime to 19.0.2

FrankReh committed:
bump wasmtime to 20.0.2

This involved changing the `WasmValue` trait to have its own `kind()`
method rather than relying on the trait to provide a ty() method that
could then have a kind() call made on its return value. With wasmtime
20.0.2, there didn't seem an easy way to create the component::Type from
a component::Val any longer, and even if there were, it would have
made for extra unnecessary work as the `writer` only really wanted the
kind anyway.

FrankReh committed:
bump wit-parser to 0.207.0

lann commented on src/wasmtime/component.rs:
```diff
@@ -107,7 +107,60 @@ impl WasmValue for component::Val {
+            Self::Float64(_) => component::Type::Float64,
+            Self::Char(_) => component::Type::Char,
+            Self::String(_) => component::Type::String,
+            _ => unimplemented!(),
+            /*
+            Self::List(_) => component::Type::List ...,
+            Self::Record(_) => component::Type::Record ...,
+            Self::Tuple(_) => component::Type::Tuple ...,
+            Self::Variant(_, _) => component::Type::Variant ...,
+            Self::Enum(_) => component::Type::Enum ...,
+            Self::Option(_) => component::Type::Option ...,
+            Self::Result(_) => component::Type::Result ...,
+            Self::Flags(_) => component::Type::Flags ...,
+            Self::Resource(_) => WasmTypeKind::Unsupported,
+             */
```
This is one of the things I was expecting to need a trait redesign to handle.

FrankReh replied:
So the tests don't run into the unimplemented cases. The `write` call uses the new `kind` method directly.

Maybe the idea of going to a Type can be changed and the logic can go directly to a Kind. That would be a more dramatic change.

Do you want to hold off reviewing the other changes until this is cleaned up? I drew inspiration from the original having used unimplemented for the more complicated types too. And do you want to do this now or is there time for me to try and figure out some more things?

Something that could help me if you think there is benefit in me trying to resolve this one, are there other unit tests you would like to seen work that aren't there yet?

lann replied:
Sorry I haven't had a chance to look at exactly what is going on here; I'll try to find time tomorrow.

FrankReh replied:
When you do look, notice that the use of val.ty().kind() has been replaced with val.kind() in two places - I think that is the interesting part. So the ty() trait method wasn't in play in the two critical areas.

lann replied:
OK, I have a better grasp on the problem now; here's what I'm thinking:

First, replace `WasmValue::ty` with `::kind`. As you note, some uses of `ty` really only need the kind.

The remaining uses of `ty` are for type checking, which this PR works around by relaxing those checks to only validate `kind`. I'd like to get that type checking back by adding a new method to `WasmValue`: `fn is(&self, ty: &Self::Type) -> bool`. I believe it should be relatively straightforward to implement this for `component::Val` though it may take a fair bit of code to handle all the composite types.

Does that sound reasonable? If you are willing to attempt the above in this PR that would be great, otherwise I'm fine merging this as-is and will probably try to make the other changes before the next release.

FrankReh replied:
> The remaining uses of ty are for type checking, which this PR works around by relaxing those checks to only validate kind. 

While I don't know *exactly* what you mean that some checks are relaxed. If you mean some calls to `ty` for a component::Val would result in a panic with the current PR commits, I agree. I wasn't sure if they were important to handle because I didn't see a unit test fail.

Would it take new unit tests to show them fail or some other API for the wasm-wave library itself that isn't fleshed out yet?

I would rather take on this next step if I could see it actually passing a test as a result.

So I can start trying to figure out the next step from your guidance, but if the current PR doesn't make things worse, I would prefer this go in, and other improvements go into a new PR. Makes understanding and possible git bisections later easier.

lann replied:
> I don't know exactly what you mean that some checks are relaxed. 

https://github.com/lann/wasm-wave/pull/34/files#r1601813873

lann replied:
The test coverage you're looking for is missing because the test cases can't be constructed just from the wasmtime API; you have to load an actual wasm file and reflect on its types to get the relevant test fixtures.

lann replied:
...though immediately after typing that I remembered that I implemented exactly that test fixture for `tests/wasmtime.rs`, so I probably can just expand coverage there.

lann replied:
Minimal test case expansion which passes on `main` and fails on this PR: https://github.com/lann/wasm-wave/pull/35

lann replied:
I'd still be fine merging this PR as-is if you prefer, we'd just need to coordinate on adding that test coverage PR if you want to work on the followup.

FrankReh replied:
Thanks for getting back to me on this. I've been following the idea that a bad value for a Flags type was being caught with wasmtime 18 and now it isn't, at least the way I had changed things.

And sure enough, a test based on the tests/wasmtime/test_round_trips_unchanged does fail with I give it a garbage flag name. I'm in the process of figuring out how the wasmtime library knew that was not valid.

FrankReh replied:
Okay. I found the wasmtime 20 library has a 'names()' method on the Flags type. So it would be straightforward to perform the check manually in that make_flags method.

Two questions. Should the error message have both the invalid string and the possible legal ones? I don't think the original gave that much information. And I wonder that the wasmtime library doesn't provide the validation check any longer. Does its author want each embedder to code their own checks?

---

Maybe that's what the comment from 3 months ago was referring to. The wasmtime library doesn't catch errors when things are created, only when they are attempted to be passed to the component (if I'm remembering that correctly).

lann replied:
Most wasmtime embedders use [`bindgen!`](https://docs.rs/wasmtime/20.0.2/wasmtime/component/macro.bindgen.html) which enforces types through code generation; they don't use this Val interface at all. I believe the API change allowed the implementation to be simplified quite a bit at a (small) expense to weird crates like this one :slightly_smiling_face:

lann replied:
Sorry, realized I didn't respond to one of your questions:

> Should the error message have both the invalid string and the possible legal ones? I don't think the original gave that much information.

I'm fine with either in this PR; nicer errors can always be added later.

lann commented on src/wasmtime/core.rs:
```diff
@@ -31,7 +32,23 @@ impl WasmValue for wasmtime::Val {
 
     fn ty(&self) -> Self::Type {
-        self.ty()
+        let config = wasmtime::Config::new();
+        let engine = wasmtime::Engine::new(&config).unwrap();
+        let store = wasmtime::Store::new(&engine, ());
+        self.ty(store)
```
I think this can be handled with a `match self`. There are `const`s for ref types (which I haven't really wrapped my head around yet), e.g. https://docs.rs/wasmtime/20.0.2/wasmtime/struct.RefType.html#associatedconstant.EXTERNREF

FrankReh replied:
Nice. At least the names match up and they are constants. See the next commit.

FrankReh committed:
cleanup ty in wasmtime/core

lann commented on src/wasmtime/component.rs:
```diff
@@ -140,60 +193,61 @@ impl WasmValue for component::Val {
+        Ok(Self::Variant(case.to_string(), val.map(Box::new)))
     }
     fn make_enum(ty: &Self::Type, case: &str) -> Result<Self, WasmValueError> {
-        ty.unwrap_enum()
-            .new_val(case)
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Enum)?;
+        Ok(Self::Enum(case.to_string()))
     }
     fn make_option(ty: &Self::Type, val: Option<Self>) -> Result<Self, WasmValueError> {
-        ty.unwrap_option()
-            .new_val(val)
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Option)?;
+        Ok(Self::Option(val.map(Box::new)))
     }
     fn make_result(
         ty: &Self::Type,
         val: Result<Option<Self>, Option<Self>>,
     ) -> Result<Self, WasmValueError> {
-        ty.unwrap_result()
-            .new_val(val)
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Result)?;
+        Ok(match val {
+            Ok(val) => Self::Result(Ok(val.map(Box::new))),
+            Err(val) => Self::Result(Err(val.map(Box::new))),
+        })
     }
     fn make_flags<'a>(
         ty: &Self::Type,
         names: impl IntoIterator<Item = &'a str>,
     ) -> Result<Self, WasmValueError> {
-        ty.unwrap_flags()
-            .new_val(&names.into_iter().collect::<Vec<_>>())
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Flags)?;
+        Ok(Self::Flags(
+            names.into_iter().map(|n| n.to_string()).collect(),
+        ))
```
Here is an example of the type checking being relaxed: [`Flags::new_val`](https://docs.rs/wasmtime/19.0.2/wasmtime/component/types/struct.Flags.html#method.new_val) which was called at 195 would validate that `names` matched the defined flag names. The same applies to many of the `make_*` diffs above.

FrankReh replied:
Oh, I was totally on the wrong track about what you were referring to.

Thanks for pointing me at this. These kinds of changes/regressions I would want to correct before the PR lands.

I can work backwards from a stack trace to see when these are being called when the crate is tested and see how more tests would have caught the regression.

FrankReh replied:
Interesting. I see the merge of the make_list test. That's confirming the composite types match up.

Wonder if there is a way to make the inputs and outputs all strings, like the other tests. Or will these require coded calls to make_* for any valid and invalid combinations. Guess they can still be put into a table of inputs and expected type results.

FrankReh replied:
How about added checks along these lines? (Note the recursion in the second function.)
```rust
    fn make_list(
        ty: &Self::Type,
        vals: impl IntoIterator<Item = Self>,
    ) -> Result<Self, WasmValueError> {
        ensure_type_kind(ty, WasmTypeKind::List)?;
        let component::Type::List(list) = ty else {
            // TBD this could be where the error is returned rather
            // than having to make the call to ensure_type_kind.
            panic!();
        };
        let elem_type = list.ty();
        vals.into_iter()
            .map(|val| {
                is_value_type(&val, &elem_type)?;
                Ok(val)
            })
            .collect::<Result<Vec<_>, WasmValueError>>()
            .map(Self::List)
    }
```
and
```rust
fn is_value_type(val: &component::Val, ty: &component::Type) -> Result<(), WasmValueError> {
    if val.kind() != ty.kind() {
        return Err(WasmValueError::wrong_value_type(ty, val));
    }
    match val {
        component::Val::List(vals) => {
            let component::Type::List(list) = ty else {
                unreachable!(); // kind to kind check should make this unreachable
            };
            let elem_type = list.ty();
            for val in vals {
                is_value_type(val, &elem_type)?;
            }
        },
        component::Val::Record(_vec_string_val) => todo!(),
        component::Val::Tuple(_vec_val_) => todo!(),
        component::Val::Variant(_string, _option_box_val__) => todo!(),
        component::Val::Enum(_string) => todo!(),
        component::Val::Option(_option_box_val__) => todo!(),
        component::Val::Result(_result_option_box_val_option_box_val) => todo!(),
        component::Val::Flags(_vec_string_) => todo!(),
        component::Val::Resource(_resourceany) => todo!(),

        // Any leaf variant type was already tested with the kind comparisons above.
        _ => {},
    }
    Ok(())
}
```
I don't like the almost duplication of code, but in the first case, I wanted to collect the values and in the second case I didn't. Open to ideas...

FrankReh replied:
I just had a thought, it could save a lot of code if the make_list creates the List but before returning it, checks that it ends up being a matching type. If the time spent and heap allocations made for the failure case is acceptable in this scenario.

FrankReh replied:
I haven't added any tests. I can understand why there weren't tests for this verify issue before because the verify was being done by the `wasmtime` library, not this one. Now that this is responsible for verifying variables are being created correctly, some more negative tests may be in order.

FrankReh committed:
Merge branch 'main' into update-wasmtime-dependency-from-18

FrankReh committed:
verify values being made against expected types

Add a function that confirms a composite component value that is
constructed from component value parts conforms to the defined type.

This verify step used to be performed by the wasmtime library when a
component type was asked to build a component value. The v20.0.0 library
was changed to have a simpler type system on the surface as the
`Resource` handling would require a `store` from the embedder.

So far, this WAVE library punts on trying to handle any `Resource`
variable so the checks made between variables and types are still
straight forward and no `store` is required.

lann commented on src/wasmtime/component.rs:
```diff
@@ -205,40 +272,166 @@ impl WasmValue for component::Val {
+
+    match val {
+        component::Val::List(vals) => {
+            let component::Type::List(list_handle) = ty else {
+                unreachable!(); // kind to kind check above should make this unreachable
+            };
```
Wasmtime conveniently provides [helpers](https://docs.rs/wasmtime/20.0.2/wasmtime/component/enum.Type.html#method.unwrap_list) for these assertions:

```suggestion
            let list_handle = ty.unwrap_list();
```

FrankReh committed:
switch to using wasmtime's unwrap_* methods

lann commented on src/wasmtime/component.rs:
```diff
@@ -107,7 +107,60 @@ impl WasmValue for component::Val {
+            Self::Float64(_) => component::Type::Float64,
+            Self::Char(_) => component::Type::Char,
+            Self::String(_) => component::Type::String,
+            _ => unimplemented!(),
```
I think it would be better to remove this method from the trait entirely rather than allow it to panic. Would that work with the changes in this PR?

FrankReh committed:
remove the `ty` trait method

This shows how it can be done.

A few TODOs left to finish, but they are doable following
the model of the ValueEnum::List case.

FrankReh commented:
Well, in addition to wondering if the direction of the last commit suits you. I was wondering if you wanted to revisit the `Value::make_*` functions in general. They are in src/value/mod.rs. I can't find that they are used except from tests, with the exception of three: Value::make_list, Value::make_option, and Value::make_result, used in src/value/convert.rs, in supporting `From` traits. So arrays, vectors, options, and results, can be converted into `Value` types.

But they all generate their `Type` information automatically, stored as a local variable `ty`, which is then passed to the `make_*` function. So the type isn't going to be compared to anything that has come from a WIT definition.

And what would trigger creation of `Value` records, variants, enums or flags.

Also wondering if the `ty` field is needed in structs in src/value/mod.rs, unless maybe a value of this type could be created with an empty vec and then operations were planned that would allow adding to the list later?
```rust
#[derive(Debug, Clone, PartialEq)]
#[doc(hidden)]
pub struct List {
    ty: Arc<ListType>,
    elements: Vec<Value>,
}
```

I know for sure I don't see all the ways this wasm-wave library was planned to be used in the future, so its current form is just a stepping stone.

Thanks.

FrankReh commented:
@lann What should we do? Want me to drop this line?

lann commented:
Ah sorry, I lost track of your comment over the weekend.

> wondering if the direction of the last commit suits you

Yes, that looks good to me.

> I was wondering if you wanted to revisit the Value::make_* functions in general.

These are used here, which is how the parser makes typed values: https://github.com/lann/wasm-wave/blob/e69b66e10383295debd16ce4bd031620fc5e1c22/src/ast.rs#L207-L236

FrankReh committed:
wasmtime 21.0.1, wit-parser 0.208.1

FrankReh commented:
Still regarding the `value` module.

I understand that for the unit tests, the types represented by the `ty` variable passed through the call chain would come from the unit test functions themselves. But that isn't helping me understand where they would come from when this is used as a library. If a WIT file or files were being used, wouldn't the library be able to use the wasmtime core or component types?

I don't yet see the need for duplicating the logic behind building variables and checking types for the concrete wasmtime core and component branches, with the library proprietary `value::{Value, Type}` variables and types.

You can always take this over if you think that is easier than getting me to see this. Thanks.

lann commented:
> wouldn't the library be able to use the wasmtime core or component types

That may be possible with the wasmtime 20 val type (it wasn't before), but there are some use cases that don't otherwise need to use wasmtime and it would represent a very heavyweight dependency just to get these value types. If these types could be moved into a 3rd party crate to be a shared dependency I think this would be feasible, but the details of resource types make that difficult, currently.

> You can always take this over if you think that is easier than getting me to see this.

I'm happy to if you'd prefer. What is the status of this PR as-is? Is it complete enough to merge?

FrankReh committed:
clippy cleanup

FrankReh committed:
verify List's builtin type

FrankReh commented:
Your comment reminded me that the wit-parser and wasmtime dependencies are optional.

FrankReh committed:
value::check_type2 Tuple arm

FrankReh commented:
The last commit has the `Tuple` checking arm fleshed out. There are a lot of types being checked.

FrankReh committed:
value::check_type2 Enum arm

FrankReh committed:
value::check_type2 Flags arm

FrankReh committed:
value::check_type2 Variant arm

FrankReh committed:
value::check_type2 Option arm

FrankReh committed:
value::check_type2 Result arm

FrankReh committed:
value::check_type2 Record arm

FrankReh commented:
@lann I think now it's ready.

Down the road, perhaps the check_type and check_type2 that are called in value/mod.rs will be called at a higher spot so they are only called once for any Node that is constructed. As it stands right now, the pieces are built up from leaf nodes and their types are also checked and then as higher order nodes are built, their types are checked again, going all the way down to the leaves. It seems unnecessary but that can wait.

Also, it seems the Record type in the value module isn't as flexible as the one in the wasmtime component module. The field names are not stored so there is no chance to provide the fields in a different order and no chance to omit optional fields. But maybe that's intentional.

lann reviewed (approved):
This looks great! Thanks for sticking though all that iteration!
